File: trunk/src/cplusplus/pyasynchio/Proactor.cpp
/*!
 *  \file Proactor.cpp
 *  \brief Exploratory wrappers over ACE_Proactor. 
 *  Implementation of pyasynchio::Proactor module.
 *  \author Vladimir Sukhoy
 */

#include <pyasynchio/detail/fix_ace_warnings.hpp>
#include <ace/Proactor.h>
#include <ace/Asynch_Acceptor.h>
#include <ace/INET_Addr.h>
#include <pyasynchio/Proactor.hpp>
#include <pyasynchio/Proactor_impl.hpp>

namespace pyasynchio {

const ACE_INET_Addr Proactor::ANY_INTERFACE(static_cast<u_short>(0));

/************************************************************************/
/* Proactor facade (PIMPL)                                              */
/************************************************************************/

Proactor::Proactor()
: pimpl_(new impl())
{
}

Proactor::~Proactor() 
{
}

void Proactor::schedule_timer(const ACE_Time_Value &delay
							  , AbstractTimerHandlerPtr elapsed)
{
    pimpl_->schedule_timer(delay, elapsed);
}

void Proactor::cancel_timer(AbstractTimerHandlerPtr elapsed)
{
    pimpl_->cancel_timer(elapsed);
}

void Proactor::open_stream_accept(AbstractAcceptHandlerPtr user_accept_handler
                      , const ACE_INET_Addr &addr 
                      , size_t bytesToRead)
{
    pimpl_->open_stream_accept(user_accept_handler, addr, bytesToRead);
}

void Proactor::close_stream_accept(AbstractAcceptHandlerPtr user_accept_handler)
{
    pimpl_->close_stream_accept(user_accept_handler);
}

void Proactor::open_stream_connect(AbstractConnectHandlerPtr user_connect_handler
                        , const ACE_INET_Addr &remote
                        , const ACE_INET_Addr &local)
{
    pimpl_->open_stream_connect(user_connect_handler, remote, local);
}

void Proactor::close_stream_connect(AbstractConnectHandlerPtr user_connect_handler)
{
    pimpl_->close_stream_connect(user_connect_handler);
}

void Proactor::open_stream_write(AbstractStreamHandlerPtr user_stream_handler
                    , const buf &data
                    , const void *act
                    , int priority
                    , int signal)
{
    pimpl_->open_stream_write(user_stream_handler, data, act, priority, signal);
}

void Proactor::cancel_stream_write(AbstractStreamHandlerPtr user_stream_handler)
{
    pimpl_->cancel_stream_write(user_stream_handler);
}

void Proactor::open_stream_read(AbstractStreamHandlerPtr user_stream_handler
                    , size_t bytes
                    , const void *act
                    , int priority
                    , int signal)
{
    pimpl_->open_stream_read(user_stream_handler, bytes, act, priority, signal);
}

void Proactor::cancel_stream_read(AbstractStreamHandlerPtr user_stream_handler)
{
    pimpl_->cancel_stream_read(user_stream_handler);
}

void Proactor::handle_events(const ACE_Time_Value &delay)
{
    pimpl_->handle_events(delay);
}

void Proactor::handle_events()
{
    pimpl_->handle_events();
}

void Proactor::close_active_stream(AbstractStreamHandlerPtr user_stream_handler)
{
    pimpl_->close_active_stream(user_stream_handler);
}

} // namespace pyasynchio

File: trunk/src/python/pyasynchio/proactor.py
from _pyasynchio import Proactor as _Proactor
from more_handlers import WrapperAcceptHandler, WrapperConnectHandler
from more_handlers import WrapperStreamHandler, WrapperTimerHandler
from more_handlers import force_wrapper, find_wrapper, kill_wrapper

class Proactor(_Proactor):
    def __init__(self):
        _Proactor.__init__(self)

    def open_stream_accept(self, accept_handler, addr):
        wah = force_wrapper(WrapperAcceptHandler, accept_handler, self)
        _Proactor.open_stream_accept(self, wah, addr)

    def close_stream_accept(self, accept_handler):
        wah = find_wrapper(WrapperAcceptHandler, accept_handler, self)
        _Proactor.close_stream_accept(self, wah)
        kill_wrapper(WrapperAcceptHandler, accept_handler, self)

    def open_stream_connect(self, connect_handler, remote, local = ('',0)):
        wch = force_wrapper(WrapperConnectHandler, accept_handler, self)
        _Proactor.open_stream_connect(self, wch, remote, local)

    def close_stream_connect(self, connect_handler):
        wch = force_wrapper(WrapperConnectHandler, connect_handler, self)
        _Proactor.close_stream_connect(self, wch)
        kill_wrapper(WrapperConnectHandler, connect_handler, self)

    def open_stream_read(self, stream_handler, byte_count):
        wsh = find_wrapper(WrapperStreamHandler, stream_handler, self)
        _Proactor.open_stream_read(self, wsh, byte_count)

    def cancel_stream_read(self, stream_handler):
        wsh = find_wrapper(WrapperStreamHandler, stream_handler, self)
        _Proactor.cancel_stream_read(self, wsh)

    def open_stream_write(self, stream_handler, data):
        wsh = find_wrapper(WrapperStreamHandler, stream_handler, self)
        _Proactor.open_stream_write(self, wsh, data)

    def cancel_stream_write(self, stream_handler):
        wsh = find_wrapper(WrapperStreamHandler, stream_handler, self)
        _Proactor.cancel_stream_write(self, wsh)

    def close_active_stream(self, stream_handler):
        wsh = find_wrapper(WrapperStreamHandler, stream_handler, self)
        _Proactor.close_active_stream(self, wsh)
        kill_wrapper(WrapperStreamHandler, stream_handler, self)
    
    def schedule_timer(self, delay, timer_handler):
        wth = force_wrapper(WrapperTimerHandler, timer_handler, self)
        _Proactor.schedule_timer(self, delay, wth)

    def cancel_timer(self, timer_handler):
        wth = find_wapper(WrapperTimerHandler, timer_handler, self)
        _Proactor.cancel_timer(self, wth)
        kill_wrapper(WrapperTimerHandler, timer_handler, self)

    def handle_events(self, delay = None):
        if delay != None:
            _Proactor.handle_events(self, delay)
        else:
            _Proactor.handle_events(self)

    def handle_events_forever(self):
        _Proactor.handle_events_foreve(self)
                                          
    